### ISSUE TITLE
Add a value processor for Settings

### ISSUE THREAD

**Issue by Xemiru** [SpongePowered/Configurate#36]
The loading right now doesn't necessarily let you control what to do when Configurate loads a value. As someone who strictly tries to idiot-proof anything the end-user can screw up, this makes having to verify values read from the configuration pretty tedious. Perhaps allowing to specify a `Function` to process the value beforehand can be added to the `@Setting` annotation? This will allow the developer to manipulate the value given by the mapper and perform actions warning the user should the returned value be invalid upon loading.

`@Setting(value = "path.to.int", processor = new IntegerCheckProcessor(defaultValue))`

**Comment by ryantheleach** [comment]
Call it a validator maybe?

**Comment by zml2008** [comment]
That isn't supported in annotations. Also, value doesn't accept multiple sublevels of data -- not sure what the dots are about.

I'm thinking the way to handle this is either a Class argument for some sort of Validator that is reflectively created -- that is imo less robust though.

**Comment by ryantheleach** [comment]
What about annotating your validator class methods, for the node it should be validating?

@Validate("path.to.int")
public boolean validateInts(int node){
return node>=0;
}

Or 
@Validate("path.to.int")
public Optional<Integer> validateInts(int node, @nullable Integer default){
if(node>=0)
  return Optional.of(node);
else
  return Optional.of(default);
}

**Comment by Xemiru** [comment]
Can go with @ryantheleach's idea, but instead it'd be easier to map the `@Validate` target to the field name, since the name is already defined in the `@Setting`.

**Comment by zml2008** [comment]
I like that @Xemiru -- might have a filtering for field names or something. I'm thinking I want to add this in along with some sort of schema-like system though, so that's probably when you'll see this.

**Comment by Meronat** [comment]
I like @dualspiral's method of handling this in Neutrino. I could port that to Configurate and PR it if that's cool with you?

https://github.com/NucleusPowered/Neutrino/blob/master/src/main/java/io/github/nucleuspowered/neutrino/annotations/ProcessSetting.java
https://github.com/NucleusPowered/Neutrino/blob/master/src/main/java/io/github/nucleuspowered/neutrino/settingprocessor/SettingProcessor.java

**Comment by dualspiral** [comment]
If it's wanted, I'll port it. It needs cleanup.

**Comment by Meronat** [comment]
@dualspiral It's definitely something I'd like to see and it does seem to accomplish what the author wants in a cleaner fashion. PR it and people can provide their opinions, but I think they will be in support.

**Comment by zml2008** [comment]
As mentioned before, I can see the value in this -- I'm not sure if directly editing the ConfigurationNode is the most elegant way to do this though. Would it be possible (or beneficial to people who want this feature) to work with the value returned from the TypeSerializer instead and have the possibility to modify that?

**Comment by dualspiral** [comment]
Yes, that would be better, I just worked with my custom object mapper and plumbed it in the same way I added my other changes. It's hacky and I wanted to change it anyway, like so are many of my other custom field annotations - but they work for now, and I've got many other things that take precidence over toying about with that.

It doesn't matter where in the chain it gets modified to me - though modifying the value before it is set (on serialisation) or after it's read (on deserialisation) is better - I shouldn't really be altering the node with the object mapper when I'm deserialising from a node.
